### PR TITLE
fix asterisk encoding

### PR DESCRIPTION
**Pull request by poddm**
Resolving errors when Configuration values contain an asterisk.

```
Error: CloudStack API error 401 (CSExceptionErrorCode: 0): unable to verify user credentials and/or request signature
```

Found the following note [here](http://docs.cloudstack.apache.org/projects/archived-cloudstack-getting-started/en/latest/dev.html):

> If you have API calls which contain * (asterisk) characters, you will need to add the option “safe = ‘*’” for the URL encoding. The reason is that Python’s urllib will encode * characters by default, while CloudStack’s internal encoder does not encode them; this results in an authentication failure for your API call. In the above you would replace “urllib.quote_plus(request[k])” with “urllib.quote_plus(request[k], safe=’*’)”.

### CODE REVIEW

**Comment by mlsorensen** [comment]
Good catch @poddm - I've run into this as well. My only feedback is that maybe if the implementation were within `encodeValues`instead,  it would be easy to add some tests.

**Comment by hasan07** [comment]
This is needed since there is a requirement for some API values to contain “ * ” such as consoleproxy.url.domain otherwise an exception is thrown. Also CloudMonkey is broked for the same API.

File: cloudstack/cloudstack.go [hunk 1]
```diff
@@ -464,17 +464,24 @@ func (cs *CloudStackClient) newRawRequest(api string, post bool, params url.Valu
...
 
-	// Generate signature for API call
+	// encode parameters
 	// * Serialize parameters, URL encoding only values and sort them by key, done by encodeValues
+	// * Replace encoded asterisk (*) back to literal.
+	//    CloudStack’s internal encoder does not encode them; this results in an authentication failure for your API call.
+	//    http://docs.cloudstack.apache.org/projects/archived-cloudstack-getting-started/en/latest/dev.html
+	// 	  Not documented http://docs.cloudstack.apache.org/en/latest/developersguide/dev.html#the-cloudstack-api
+	encodedParams := encodeValues(params)
+	encodedParams = strings.Replace(encodedParams, "%2A", "*", -1)
```
[the remark this answers is not in the document]
Comment (mlsorensen) [reply cloudstack/cloudstack.go:474]: It does, you can't set a value with an asterisk via cloud monkey.

**Comment by mlsorensen** [comment]
@poddm looks like the change would need to be made in the generate.go in order to stick, since the code changed is generated.

https://github.com/apache/cloudstack-go/blob/0106fc79c6b1031ad6d7c89a157ba8a77f6c8afe/generate/generate.go#L529

File: cloudstack/cloudstack.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (rohityadavcloud) [reply cloudstack/cloudstack.go:474]: Thanks @mlsorensen, @poddm https://github.com/apache/cloudstack-cloudmonkey/blob/main/cmd/network.go can use a PR as well.

**Comment by poddm** [comment]
> 0106fc7

**Comment by poddm** [comment]
@mlsorensen, overlooked this.  I'll update the PR.

**Comment by mlsorensen** [comment]
@poddm see what you think about #52

**Comment by poddm** [comment]
lgtm -

**Comment by shwstppr** [comment]
@poddm @mlsorensen do we need this now as #52 is already merged?

**Comment by mlsorensen** [comment]
> @poddm @mlsorensen do we need this now as #52 is already merged?

I don't think so.

**Pull request closed without merging**